src/charm.py

```diff
@@ -240,27 +239,26 @@ def set_ports(self):
         for p in new_ports_to_open:
             self.unit.open_port(p.protocol, p.port)
 
-    def _ingress_scheme(self):
-        if self.is_tls_enabled():
-            return "https"
-        return "http"
-
     @property
     def self_scraping_job(self):
         """The self-monitoring scrape job."""
-        external_url = urlparse(self._external_url)
-        metrics_path = f"{external_url.path.rstrip('/')}/metrics"
-        target = (
-            f"{external_url.hostname}{':'+str(external_url.port) if external_url.port else ''}"
-        )
-        job = {
+        # We assume that scraping, especially self-monitoring, is in-cluster.
+        # This assumption is necessary because the local CA signs CSRs with FQDN as the SAN DNS.
+        # If prometheus were to scrape an ingress URL instead, it would error out with:
+        # x509: cannot validate certificate.
+        metrics_endpoint = urlparse(self._internal_url.rstrip("/") + "/metrics")
+        metrics_path = metrics_endpoint.path
+        # Render a ':port' section only if it is explicit (e.g. 9093; without a port, the port is
+        # deduced from scheme).
+        port_str = (":" + str(metrics_endpoint.port)) if metrics_endpoint.port else ""
```

Comment:
very edge case, but if the port is 0 this will misbehave. I strive to always do non-null comparisons with `is not None` just to be safe

Reply:
Nice spotting! Fixed in next commit.